### PR TITLE
Feature - Scrub tool for each repo

### PR DESCRIPTION
**Pull request by richmahn**
Adds the "Remove Sensitive Data" tool to every repo's setting page. 

See https://github.com/unfoldingWord-dev/gogs/issues/91 and https://github.com/unfoldingWord-dev/door43.org/issues/340 for the specs.

### CODE REVIEW

File: models/repo.go [hunk 1]
```diff
@@ -1903,6 +1905,128 @@ func (repos MirrorRepositoryList) LoadAttributes() error {
 	return repos.loadAttributes(x)
 }
 
+type ScrubSensativeDataOptions struct {
+	LastCommitID	string
+	CommitMessage	string
+}
+
+// SrubSenstaiveData removes emails and passwords from the manifest.json or project.json file's history.
```
Comment (ethantkoenig) [inline models/repo.go]: Spelling of `ScrubSensativeData`

File: routers/repo/setting.go [hunk 2]
```diff
@@ -288,6 +288,40 @@ func SettingsPost(ctx *context.Context, form auth.RepoSettingForm) {
 		ctx.Flash.Success(ctx.Tr("repo.settings.wiki_deletion_success"))
 		ctx.Redirect(ctx.Repo.RepoLink + "/settings")
 
+	case "scrub":
+		if !ctx.Repo.IsOwner() {
+			ctx.Error(404)
+			return
+		}
+		if repo.Name != form.RepoName {
+			ctx.RenderWithErr(ctx.Tr("form.enterred_invalid_repo_name"), tplSettingsOptions, nil)
```
Comment (ethantkoenig) [inline routers/repo/setting.go]: entered?

File: vendor/code.gitea.io/git/repo.go [hunk 3]
```diff
@@ -153,6 +153,13 @@ func Push(repoPath, remote, branch string) error {
 	return err
 }
 
+// PushFroce pushs local commits to given remote branch and forces any changes.
```
Comment (ethantkoenig) [inline vendor/code.gitea.io/git/repo.go]: Spelling of `PushForce`, "pushes"

File: vendor/code.gitea.io/git/scrub.go [hunk 4]
```diff
@@ -0,0 +1,32 @@
+package git
+
+import (
+	"os"
+	"path"
+	"fmt"
+	"os/exec"
+)
+
+// Scrub file completely removes a file from a repository's history
```
Comment (ethantkoenig) [inline vendor/code.gitea.io/git/scrub.go]: For the sake of consistency, this comment should start with `ScrubFile` (https://blog.golang.org/godoc-documenting-go-code)

File: vendor/code.gitea.io/git/repo.go [hunk 5]
```diff
@@ -153,6 +153,13 @@ func Push(repoPath, remote, branch string) error {
...
+func PushForce(repoPath, remote, branch string) error {
+	_, err := NewCommand("push", "--force", "--all", remote, branch).RunInDir(repoPath)
+	_, err = NewCommand("push", "--force", "--tags", remote, branch).RunInDir(repoPath)
```
Comment (ethantkoenig) [inline vendor/code.gitea.io/git/repo.go]: Should there be a `if err != nil { return err }` check before this line? It seems that we completely ignore the `err` value returned by the first call.

**Comment by phillip-hopper** [comment]
@richmahn: When I run it locally I'm not finding a "Scrub" button. Where should I find it. This is what I see in the Danger Zone:

image

File: models/repo.go [hunk 6]
```diff
@@ -1903,6 +1905,128 @@ func (repos MirrorRepositoryList) LoadAttributes() error {
...
+		return fmt.Errorf("DiscardLocalRepoBranchChanges [branch: master]: %v", err)
+	} else if err = repo.UpdateLocalCopyBranch("master"); err != nil {
+		return fmt.Errorf("UpdateLocalCopyBranch [branch: master]: %v", err)
+	}
+
+	dirty := false
+	dirty = repo.ScrubJsonFile("project.json") || dirty
+	dirty = repo.ScrubJsonFile("package.json") || dirty
+	dirty = repo.ScrubJsonFile("manifest.json") || dirty
+	dirty = repo.ScrubJsonFile("status.json") || dirty
```
Comment (phillip-hopper) [inline models/repo.go]: `settings.scrub_notices_1` says it will only scrub manifest.json and project.json

File: models/repo.go (hunk 1 above)
Comment (richmahn) [inline models/repo.go]: Yes, we are learning I can't spell Sensitive to save my life.....and I'm very sensative about it. 😛

**Comment by jag3773** [comment]
@phillip-hopper I'd rather not "scrub all yaml and json files in the repository root" .  Mostly I'm concerned about messing with someone else's work.

**Comment by jag3773** [comment]
@richmahn I agree with @phillip-hopper, test coverage must *increase*.

File: routers/repo/setting.go (hunk 2 above)
Comment (richmahn) [inline routers/repo/setting.go]: I didn't come up with that key for the text. It is used throughout setting.go

**Comment by richmahn** [comment]
@phillip-hopper: I was unable to figure out how to even make a test for the routers/ or models/ directory. Are the functions in the same file as my functions tested somewhere? Unable to find them.

**Comment by richmahn** [comment]
@jag3773: Didn't even know there was any test coverage...did see some for modules in Gogs, but not for the rest of the MVC aspect of Gogs...will investigate.

**Comment by richmahn** [comment]
@phillip-hopper: I had added the scrub tool stuff to the gogs-custom repo, but am now leaning more for just putting it in our fork of Gogs.....already have started modifying the main template files and it is actually easier to make sure we get changes to templates from upstream if we use the templates/ dir in Gogs. So now the scrub tool is in the gogs repo's templates/repo/settings/options.tmpl

**Comment by phillip-hopper** [comment]
@richmahn In https://github.com/unfoldingWord-dev/gogs/pull/104 there is a test for `models/hashtag.go` (hashtag_test.go).

**Comment by richmahn** [comment]
@phillip-hopper: I know, I made that test for hashtag, haha. But before now I didn't have a module for scrub, just stuff in the router and models. I actually tried to figure out a way to test from there, but that would mean I need the whole system to load a repo into an object, etc. etc. Yet on Travis nothing is set up, so I had to move code I wanted tested to not rely on Gogs, but just on regular file handling.

**Pull request merged**